core/src/main/java/org/elasticsearch/index/similarity/SimilarityService.java

```diff
@@ -122,15 +131,15 @@ public SimilarityProvider getSimilarity(String name) {
     }
 
     Similarity getDefaultSimilarity() {
-        return defaultSimilarity;
+        return defaultSimilarity.get();
     }
 
     static class PerFieldSimilarity extends PerFieldSimilarityWrapper {
 
-        private final Similarity defaultSimilarity;
+        private final SimilarityProvider defaultSimilarity;
         private final MapperService mapperService;
 
-        PerFieldSimilarity(Similarity defaultSimilarity, Similarity baseSimilarity, MapperService mapperService) {
+        PerFieldSimilarity(SimilarityProvider defaultSimilarity, Similarity baseSimilarity, MapperService mapperService) {
```

Comment:
> I think we shouldn't implement the update the way we do it here but rather keep all the provider immutable and if there is an update to any of the similarity settings we rebuild the entire set of providers. this would guarantee that we have a consistent view on all the values as well as a consistent view across all users ie one search sees one set of similarities? does this make sense?

It makes perfect sense though it would require a lot of modifications on how we handle similarity at indexing and search time. In master we build a per field similarity wrapper once for each searcher/writer. With immutable providers we would need to be able to dynamically change the similarity of the ES searcher/writer. Furthermore to ensure that a query/document gets a consistent view of the similarity we would need to switch the similarity only for new queries/documents. To do that I don't see other way than forcing a refresh of the searcher/writer for each similarity change. 
Does these sound acceptable ?
I though that this solution was too invasive so I followed the synchronized/volatile path which simplifies a lot of things but does not guarantee consistency and may impact the performance. I realize with your comment that the latter is a blocker even though the impact on performance has not been proven yet.